Warn and skip code gen for duplicate aliases

PR #2455 created a conflict with existing Cake.Incubator addin
```diff
- (2688,10): error CS0111: Type 'Submission#0' already defines a member called 'EnvironmentVariable' with the same parameter types
```

Repro script
```cake
#addin nuget:?package=Cake.Incubator&version=3.1.0
```

What we think we need to do here is track methods generated by code gen and skip & warn on any duplicates, prioritizing what's in Cake.Common.

Likely would need to add using statements to avoid errors with aliases being invoked on `ICakeContext`

--- a/src/Cake.Core/Scripting/CodeGen/MethodAliasGenerator.cs
+++ b/src/Cake.Core/Scripting/CodeGen/MethodAliasGenerator.cs
@@ -17,6 +17,8 @@ namespace Cake.Core.Scripting.CodeGen
     /// </summary>
     public static class MethodAliasGenerator
     {
+        private static readonly System.Security.Cryptography.SHA256 SHA256 = System.Security.Cryptography.SHA256.Create();
+
         /// <summary>
         /// Generates a script method alias from the specified method.
         /// The provided method must be an extension method for <see cref="ICakeContext"/>
@@ -24,7 +26,17 @@ namespace Cake.Core.Scripting.CodeGen
         /// </summary>
         /// <param name="method">The method to generate the code for.</param>
         /// <returns>The generated code.</returns>
-        public static string Generate(MethodInfo method)
+        public static string Generate(MethodInfo method) => Generate(method, out _);
+
+        /// <summary>
+        /// Generates a script method alias from the specified method.
+        /// The provided method must be an extension method for <see cref="ICakeContext"/>
+        /// and it must be decorated with the <see cref="CakeMethodAliasAttribute"/>.
+        /// </summary>
+        /// <param name="method">The method to generate the code for.</param>
+        /// <param name="hash">The hash of method signature.</param>
+        /// <returns>The generated code.</returns>
+        public static string Generate(MethodInfo method, out string hash)
         {
             if (method == null)
             {
@@ -56,6 +68,12 @@ namespace Cake.Core.Scripting.CodeGen
             {
                 GenericParameterConstraintEmitter.BuildGenericConstraints(method, builder);
             }
+
+            hash = SHA256
+                .ComputeHash(Encoding.UTF8.GetBytes(builder.ToString()))
+                .Aggregate(new StringBuilder(),
+                (sb, b) => sb.AppendFormat("{0:x2}", b),
+                sb => sb.ToString());
 
             builder.AppendLine();
             builder.Append("{");

--- a/src/Cake.Core/Scripting/CodeGen/PropertyAliasGenerator.cs
+++ b/src/Cake.Core/Scripting/CodeGen/PropertyAliasGenerator.cs
@@ -5,6 +5,7 @@
 using System;
 using System.Diagnostics;
 using System.Globalization;
+using System.Linq;
 using System.Reflection;
 using System.Runtime.CompilerServices;
 using System.Text;
@@ -17,6 +18,8 @@ namespace Cake.Core.Scripting.CodeGen
     /// </summary>
     public static class PropertyAliasGenerator
     {
+        private static readonly System.Security.Cryptography.SHA256 SHA256 = System.Security.Cryptography.SHA256.Create();
+
         /// <summary>
         /// Generates a script property alias from the specified method.
         /// The provided method must be an extension method for <see cref="ICakeContext"/>
@@ -24,7 +27,17 @@ namespace Cake.Core.Scripting.CodeGen
         /// </summary>
         /// <param name="method">The method to generate the code for.</param>
         /// <returns>The generated code.</returns>
-        public static string Generate(MethodInfo method)
+        public static string Generate(MethodInfo method) => Generate(method, out _);
+
+        /// <summary>
+        /// Generates a script property alias from the specified method.
+        /// The provided method must be an extension method for <see cref="ICakeContext"/>
+        /// and it must be decorated with the <see cref="CakePropertyAliasAttribute"/>.
+        /// </summary>
+        /// <param name="method">The method to generate the code for.</param>
+        /// <param name="hash">The hash of property signature.</param>
+        /// <returns>The generated code.</returns>
+        public static string Generate(MethodInfo method, out string hash)
         {
             if (method == null)
             {
@@ -40,8 +53,8 @@ namespace Cake.Core.Scripting.CodeGen
 
             // Generate code.
             return attribute.Cache
-                ? GenerateCachedCode(method)
-                    : GenerateCode(method);
+                ? GenerateCachedCode(method, out hash)
+                    : GenerateCode(method, out hash);
         }
 
         private static void ValidateMethod(MethodInfo method)
@@ -98,15 +111,12 @@ namespace Cake.Core.Scripting.CodeGen
             }
         }
 
-        private static string GenerateCode(MethodInfo method)
+        private static string GenerateCode(MethodInfo method, out string hash)
         {
             var builder = new StringBuilder();
 
-            builder.Append("public ");
-            builder.Append(GetReturnType(method));
-            builder.Append(" ");
-            builder.Append(method.Name);
-            builder.AppendLine();
+            hash = GenerateCommonInitalCode(method, ref builder);
+
             builder.Append("{");
             builder.AppendLine();
             builder.Append("    get");
@@ -150,7 +160,25 @@ namespace Cake.Core.Scripting.CodeGen
             return builder.ToString();
         }
 
-        private static string GenerateCachedCode(MethodInfo method)
+        private static string GenerateCommonInitalCode(MethodInfo method, ref StringBuilder builder)
+        {
+            string hash;
+            var curPos = builder.Length;
+            builder.Append("public ");
+            builder.Append(GetReturnType(method));
+            builder.Append(" ");
+            builder.Append(method.Name);
+            builder.AppendLine();
+
+            hash = SHA256
+                .ComputeHash(Encoding.UTF8.GetBytes(builder.ToString(curPos, builder.Length - curPos)))
+                .Aggregate(new StringBuilder(),
+                (sb, b) => sb.AppendFormat("{0:x2}", b),
+                sb => sb.ToString());
+            return hash;
+        }
+
+        private static string GenerateCachedCode(MethodInfo method, out string hash)
         {
             var builder = new StringBuilder();
 
@@ -160,7 +188,7 @@ namespace Cake.Core.Scripting.CodeGen
             {
                 if (obsolete.IsError)
                 {
-                    return GenerateCode(method);
+                    return GenerateCode(method, out hash);
                 }
             }
 
@@ -177,11 +205,7 @@ namespace Cake.Core.Scripting.CodeGen
             builder.AppendLine();
 
             // Property
-            builder.Append("public ");
-            builder.Append(GetReturnType(method));
-            builder.Append(" ");
-            builder.Append(method.Name);
-            builder.AppendLine();
+            hash = GenerateCommonInitalCode(method, ref builder);
             builder.Append("{");
             builder.AppendLine();
             builder.AppendLine("    get");

--- a/src/Cake.Core/Scripting/Script.cs
+++ b/src/Cake.Core/Scripting/Script.cs
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
 using System.Collections.Generic;
 using System.Linq;
 
@@ -12,18 +13,16 @@ namespace Cake.Core.Scripting
     /// </summary>
     public sealed class Script
     {
-        private readonly List<string> _namespaces;
-        private readonly List<string> _lines;
-        private readonly List<ScriptAlias> _aliases;
-        private readonly List<string> _usingAliasDirectives;
-        private readonly List<string> _usingStaticDirectives;
-        private readonly List<string> _defines;
-
         /// <summary>
         /// Gets the namespaces imported via the <c>using</c> statement.
         /// </summary>
         /// <value>The namespaces.</value>
-        public IReadOnlyList<string> Namespaces => _namespaces;
+        public IReadOnlyList<string> Namespaces { get; }
+
+        /// <summary>
+        /// Gets the namespaces flagged to be excluded by code generation and affected aliases.
+        /// </summary>
+        public IDictionary<string, IList<string>> ExcludedNamespaces { get;  }
 
         /// <summary>
         /// Gets the script lines.
@@ -31,31 +30,31 @@ namespace Cake.Core.Scripting
         /// <value>
         /// The lines.
         /// </value>
-        public IReadOnlyList<string> Lines => _lines;
+        public IReadOnlyList<string> Lines { get; }
 
         /// <summary>
         /// Gets the aliases.
         /// </summary>
         /// <value>The aliases.</value>
-        public IReadOnlyList<ScriptAlias> Aliases => _aliases;
+        public IReadOnlyList<ScriptAlias> Aliases { get; }
 
         /// <summary>
         /// Gets the using alias directives.
         /// </summary>
         /// <value>The using alias directives.</value>
-        public IReadOnlyList<string> UsingAliasDirectives => _usingAliasDirectives;
+        public IReadOnlyList<string> UsingAliasDirectives { get; }
 
         /// <summary>
         /// Gets the using static directives.
         /// </summary>
         /// <value>The using static directives.</value>
-        public IReadOnlyList<string> UsingStaticDirectives => _usingStaticDirectives;
+        public IReadOnlyList<string> UsingStaticDirectives { get; }
 
         /// <summary>
         /// Gets the defines.
         /// </summary>
         /// <value>The defines.</value>
-        public IReadOnlyList<string> Defines => _defines;
+        public IReadOnlyList<string> Defines { get; }
 
         /// <summary>
         /// Initializes a new instance of the <see cref="Script" /> class.
@@ -74,12 +73,13 @@ namespace Cake.Core.Scripting
             IEnumerable<string> usingStaticDirectives,
             IEnumerable<string> defines)
         {
-            _namespaces = new List<string>(namespaces ?? Enumerable.Empty<string>());
-            _lines = new List<string>(lines ?? Enumerable.Empty<string>());
-            _aliases = new List<ScriptAlias>(aliases ?? Enumerable.Empty<ScriptAlias>());
-            _usingAliasDirectives = new List<string>(usingAliasDirectives ?? Enumerable.Empty<string>());
-            _usingStaticDirectives = new List<string>(usingStaticDirectives ?? Enumerable.Empty<string>());
-            _defines = new List<string>(defines ?? Enumerable.Empty<string>());
+            Namespaces = new List<string>(namespaces ?? Enumerable.Empty<string>());
+            Lines = new List<string>(lines ?? Enumerable.Empty<string>());
+            Aliases = new List<ScriptAlias>(aliases ?? Enumerable.Empty<ScriptAlias>());
+            UsingAliasDirectives = new List<string>(usingAliasDirectives ?? Enumerable.Empty<string>());
+            UsingStaticDirectives = new List<string>(usingStaticDirectives ?? Enumerable.Empty<string>());
+            Defines = new List<string>(defines ?? Enumerable.Empty<string>());
+            ExcludedNamespaces = new Dictionary<string, IList<string>>(StringComparer.Ordinal);
         }
     }
 }

--- a/src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs
+++ b/src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs
@@ -22,14 +22,37 @@ namespace Cake.Scripting.Roslyn
 
         private static string GetAliasCode(Script context)
         {
-            var result = new List<string>();
+            var result = new Dictionary<string, string>();
             foreach (var alias in context.Aliases)
             {
-                result.Add(alias.Type == ScriptAliasType.Method
-                    ? MethodAliasGenerator.Generate(alias.Method)
-                    : PropertyAliasGenerator.Generate(alias.Method));
+                string hash, code = alias.Type == ScriptAliasType.Method
+                    ? MethodAliasGenerator.Generate(alias.Method, out hash)
+                    : PropertyAliasGenerator.Generate(alias.Method, out hash);
+
+                string @namespace = alias.Method.DeclaringType.Namespace ?? "@null";
+                if (result.ContainsKey(hash))
+                {
+                    var message = $"{alias.Type.ToString().ToLowerInvariant()} \"{alias.Name}\" excluded from code generation and will need to be fully qualified on ICakeContext.";
+                    if (context.ExcludedNamespaces.ContainsKey(@namespace))
+                    {
+                        context.ExcludedNamespaces[@namespace].Add(message);
+                    }
+                    else
+                    {
+                        context.ExcludedNamespaces.Add(@namespace,
+                            new List<string>() { message });
+                    }
+                    continue;
+                }
+                else if (context.ExcludedNamespaces.ContainsKey(@namespace))
+                {
+                    var message = $"{alias.Type.ToString().ToLowerInvariant()} \"{alias.Name}\" was included in code generation, but will need to be fully qualified on ICakeContext.";
+                    context.ExcludedNamespaces[@namespace].Add(message);
+                }
+
+                result.Add(hash, code);
             }
-            return string.Join("\r\n", result);
+            return string.Join("\r\n", result.Values);
         }
     }
 }

--- a/src/Cake/Scripting/Roslyn/RoslynScriptSession.cs
+++ b/src/Cake/Scripting/Roslyn/RoslynScriptSession.cs
@@ -82,9 +82,16 @@ namespace Cake.Scripting.Roslyn
             var generator = new RoslynCodeGenerator();
             var code = generator.Generate(script);
 
+            // Warn about any code generation excluded namespaces
+            foreach (var @namespace in script.ExcludedNamespaces)
+            {
+                _log.Warning("Namespace {0} excluded by code generation, affected methods:\r\n\t{1}",
+                    @namespace.Key, string.Join("\r\n\t", @namespace.Value));
+            }
+
             // Create the script options dynamically.
             var options = Microsoft.CodeAnalysis.Scripting.ScriptOptions.Default
-                .AddImports(Namespaces)
+                .AddImports(Namespaces.Except(script.ExcludedNamespaces.Keys))
                 .AddReferences(References)
                 .AddReferences(ReferencePaths.Select(r => r.FullPath))
                 .WithEmitDebugInformation(_options.PerformDebug)

--- a/tests/integration/Cake.Core/Scripting/AddinDirective.cake
+++ b/tests/integration/Cake.Core/Scripting/AddinDirective.cake
@@ -1,3 +1,5 @@
+#addin nuget:?package=Cake.Incubator&version=4.0.0
+
 Task("Cake.Core.Scripting.AddinDirective.LoadNetStandardAddin")
     .Does(() =>
 {
@@ -12,15 +14,24 @@ Task("Cake.Core.Scripting.AddinDirective.LoadNetStandardAddin")
     ";
 
     CakeExecuteExpression(script,
-    	new CakeSettings {
-    		EnvironmentVariables = new Dictionary<string, string>{
-    			{"CAKE_PATHS_ADDINS", $"{Paths.Temp}/tools/Addins"},
-    			{"CAKE_PATHS_TOOLS", $"{Paths.Temp}/tools"},
-    			{"CAKE_PATHS_MODULES", $"{Paths.Temp}/tools/Modules"}
-    		}});
+        new CakeSettings {
+            EnvironmentVariables = new Dictionary<string, string>{
+                {"CAKE_PATHS_ADDINS", $"{Paths.Temp}/tools/Addins"},
+                {"CAKE_PATHS_TOOLS", $"{Paths.Temp}/tools"},
+                {"CAKE_PATHS_MODULES", $"{Paths.Temp}/tools/Modules"},
+            },
+            Verbosity = Context.Log.Verbosity
+        });
+});
+
+Task("Cake.Core.Scripting.AddinDirective.CallDuplicatedMethod")
+    .Does(context =>
+{
+    var result = context.EnvironmentVariable("CAKE_DOES_ROCK", true);
 });
 
 //////////////////////////////////////////////////////////////////////////////
 
 Task("Cake.Core.Scripting.AddinDirective")
-    .IsDependentOn("Cake.Core.Scripting.AddinDirective.LoadNetStandardAddin");
+    .IsDependentOn("Cake.Core.Scripting.AddinDirective.LoadNetStandardAddin")
+    .IsDependentOn("Cake.Core.Scripting.AddinDirective.CallDuplicatedMethod");

